docs/design/1443-dynamic-pipelines.md

```diff
@@ -0,0 +1,252 @@
+}
+```
+
+<details>
+  <summary>Example</summary>
+
+```alloy
+// declare.dynamic "maps" each target to a sub-pipeline.
+// Each sub-pipeline has 1 exporter, 1 relabel, and 1 scraper.
+// Internally, maybe one way this can be done via serializing the pipeline to a string and then importing it as a module?
+declare.dynamic "redis_exporter" {
+  argument "input_targets" {
+    optional = false
+    comment = "We will create a sub-pipeline for each target in input_targets."
+  }
+
+  argument "output_metrics" {
+    optional = false
+    comment = "All the metrics gathered from all pipelines."
+  }
+
+  // "id" is a special identifier for every "sub-pipeline".
+  // The number of "sub-pipelines" is equal to len(input_targets).
+  prometheus.exporter.redis id {
+    redis_addr = input_targets["__address__"]
+  }
+
+  discovery.relabel id {
+    targets = prometheus.exporter.redis[id].targets
+    // Add a label which comes from the discovery component.
+    rule {
+      target_label = "filepath"
+      // __meta_filepath comes from discovery.file
+      replacement  = input_targets["__meta_filepath"]
+    }
+  }
+
+  prometheus.scrape id {
+    targets = prometheus.exporter.redis[id].targets
+    forward_to = output_metrics
+  }
+
+}
+discovery.file "default" {
+  files = ["/Users/batman/Desktop/redis_addresses.yaml"]
+}
+
+declare.dynamic.redis_exporter "default" {
+  input_targets = discovery.file.default.targets
+  output_metrics = [prometheus.remote_write.mimir.receiver]
+}
+
+prometheus.remote_write "mimir" {
+  endpoint {
+    url = "https://prometheus-prod-05-gb-south-0.grafana.net/api/prom/push"
+    basic_auth {
+      username = ""
+      password = ""
+    }
+  }
+}
+```
+
+</details>
+
+Pros:
+* Looks more like a component than a `foreach` block.
+* Flexible number of inputs and outputs.
+
+Cons:
+* A name such as `declare.dynamic` doesn't sound as familiar to most people than `foreach`.
+* It may not be practical to implement this in a way that there's more than one possible input collection.
+  * How can we limit users to having just one collection?
+* Having another variant of the `declare` block can feel complex.
+  Can we just add this functionality to the normal `declare` block, so that we can avoid having a `declare.dynamic` block?
+
+## Proposal 3: Do nothing
+
+It is customary to always include a "do nothing" proposal, in order to evaluate if the work is really required.
+
+Pros:
+* No effort required.
+* Alloy's syntax is simpler since we're not adding any new types of blocks.
+
+Cons:
+* Not possible to integrate most `prometheus.exporter` components with the `discovery` ones.
+
+## Unknowns
+
+We should find answers to the unknowns below before this proposal is accepted:
+
+* Will the solution only work for `list()`? What about `map()`?
+  * If we go with a `foreach`, we could have a `key` attribute in addition to the `var` one. That way we can also access the key. The `key` attribute can be a no-op if `collection` is a map?
```

Comment:
Alternatively, a map of `{k1: v1, k2: v2}` can be translated into list of entries: `[{key: k1, value: v1}, {key: k2, value: v2}]`.